File: src/api/foreground.ts
/**
 * 前台检测 API
 *
 * 用于检测用户当前前台窗口是否为娱乐类应用，触发惩罚机制。
 * 对应 Electron 旧版 foregroundInspection + ipc-foreground.js。
 *
 * 命令命名（Rust 端 snake_case）：
 * - foreground_is_ready
 * - foreground_start / foreground_stop
 * - foreground_get_status
 * - foreground_set_api_key(api_key)
 * - foreground_add_whitelist(keyword) / foreground_add_blacklist(keyword)
 * - foreground_mark_history_not(window_title)
 * - foreground_move_blacklist_to_whitelist(keyword)
 * - foreground_get_config
 *
 * 事件名（与旧 Electron 版一致，kebab-case）：
 * - 'foreground-ready'
 * - 'foreground-api-key-invalid'
 * - 'foreground-entertainment-detected'
 * - 'foreground-status'
 * - 'foreground-error'
 */
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn, type EventCallback } from "@tauri-apps/api/event";

// ===== 类型定义 =====

/** 前台检测结果（与 Rust `foreground_inspection::DetectionResult` 对应） */
export interface DetectionResult {
  /** 前台窗口标题 */
  windowTitle: string;
  /** 是否被判定为娱乐 */
  isEntertainment: boolean;
  /** 判定来源：'blacklist' | 'whitelist' | 'ai' | 'history' */
  source: string;
  /** 命中的关键词（如有） */
  keyword: string;
}

/** 前台检测运行状态 */
export interface ForegroundStatus {
  /** 是否正在运行检测 */
  running: boolean;
  /** 是否已配置 API Key */
  hasApiKey: boolean;
  /** 最近一次检测时间（毫秒时间戳） */
  lastCheckAt?: number;
  /** 最近一次检测结果 */
  lastResult?: DetectionResult;
}

// ===== 命令调用 =====

/**
 * 查询前台检测是否就绪（模块已加载）。
 * 后端：`foreground_is_ready() -> bool`
 */
export function foregroundIsReady(): Promise<boolean> {
  return invoke<boolean>("foreground_is_ready");
}

/**
 * 启动前台检测循环。
 * 后端：`foreground_start()`
 */
export function foregroundStart(): Promise<void> {
  return invoke<void>("foreground_start");
}

/**
 * 停止前台检测循环。
 * 后端：`foreground_stop()`
 */
export function foregroundStop(): Promise<void> {
  return invoke<void>("foreground_stop");
}

/**
 * 主动查询当前前台检测状态。
 * 后端：`foreground_get_status() -> Option<ForegroundStatus>`
 */
export function foregroundGetStatus(): Promise<ForegroundStatus | null> {
  return invoke<ForegroundStatus | null>("foreground_get_status");
}

/**
 * 设置 DeepSeek API Key（用于 AI 判断娱乐性）。
 * 后端：`foreground_set_api_key(api_key: String)`
 */
export function foregroundSetApiKey(key: string): Promise<void> {
  return invoke<void>("foreground_set_api_key", { apiKey: key });
}

// ===== 名单管理 =====

/** 名单配置（与 Rust `foreground_inspection::ListConfig` 对应） */
export interface ForegroundListConfig {
  /** 用户自定义白名单关键词 */
  whitelist: string[];
  /** 用户自定义黑名单关键词 */
  blacklist: string[];
  /** AI 判断历史：窗口标题 → 是否娱乐 */
  history: Record<string, boolean>;
}

/**
 * 添加关键词到白名单（命中 → 非娱乐）。返回是否真正新增（重复/空关键词返回 false）。
 * 后端：`foreground_add_whitelist(keyword) -> bool`
 */
export function foregroundAddWhitelist(keyword: string): Promise<boolean> {
  return invoke<boolean>("foreground_add_whitelist", { keyword });
}

/**
 * 添加关键词到黑名单（命中 → 娱乐）。返回是否真正新增。
 * 后端：`foreground_add_blacklist(keyword) -> bool`
 */
export function foregroundAddBlacklist(keyword: string): Promise<boolean> {
  return invoke<boolean>("foreground_add_blacklist", { keyword });
}

/**
 * 将历史记录中的窗口标题标记为"不是娱乐"。
 * ForegroundWarning"不是娱乐"按钮：source 为 history/ai 时调用。
 * 后端：`foreground_mark_history_not(window_title) -> bool`
 */
export function foregroundMarkHistoryNot(windowTitle: string): Promise<boolean> {
  return invoke<boolean>("foreground_mark_history_not", { windowTitle });
}

/**
 * 把黑名单关键词移到白名单（误判纠正）。
 * ForegroundWarning"不是娱乐"按钮：source 为 blacklist 时调用。
 * 后端：`foreground_move_blacklist_to_whitelist(keyword) -> bool`
 */
export function foregroundMoveBlacklistToWhitelist(
  keyword: string,
): Promise<boolean> {
  return invoke<boolean>("foreground_move_blacklist_to_whitelist", { keyword });
}

/**
 * 获取当前名单配置（用户白/黑名单 + 历史记录）。
 * 后端：`foreground_get_config() -> ForegroundListConfig`
 */
export function foregroundGetConfig(): Promise<ForegroundListConfig> {
  return invoke<ForegroundListConfig>("foreground_get_config");
}

// ===== 事件监听 =====

/** 事件名常量 */
export const FOREGROUND_EVENTS = {
  /** 前台检测就绪 */
  ready: "foreground-ready",
  /** API Key 失效 */
  apiKeyInvalid: "foreground-api-key-invalid",
  /** 检测到娱乐类应用 */
  entertainmentDetected: "foreground-entertainment-detected",
  /** 状态更新 */
  status: "foreground-status",
  /** 检测出错 */
  error: "foreground-error",
} as const;

/**
 * 监听 `foreground-ready` 事件。
 * 返回 Promise<UnlistenFn>，调用方需在适当时机调用以取消监听。
 */
export function onForegroundReady(
  handler: EventCallback<unknown>,
): Promise<UnlistenFn> {
  return listen(FOREGROUND_EVENTS.ready, handler);
}

/**
 * 监听 `foreground-api-key-invalid` 事件。
 */
export function onForegroundApiKeyInvalid(
  handler: EventCallback<unknown>,
): Promise<UnlistenFn> {
  return listen(FOREGROUND_EVENTS.apiKeyInvalid, handler);
}

/**
 * 监听 `foreground-entertainment-detected` 事件（payload 为 DetectionResult）。
 */
export function onForegroundEntertainmentDetected(
  handler: EventCallback<DetectionResult>,
): Promise<UnlistenFn> {
  return listen<DetectionResult>(
    FOREGROUND_EVENTS.entertainmentDetected,
    handler,
  );
}

/**
 * 监听 `foreground-status` 事件（payload 为 ForegroundStatus）。
 */
export function onForegroundStatus(
  handler: EventCallback<ForegroundStatus>,
): Promise<UnlistenFn> {
  return listen<ForegroundStatus>(FOREGROUND_EVENTS.status, handler);
}

/**
 * 监听 `foreground-error` 事件（payload 为错误信息）。
 */
export function onForegroundError(
  handler: EventCallback<string>,
): Promise<UnlistenFn> {
  return listen<string>(FOREGROUND_EVENTS.error, handler);
}

File: src-tauri/src/commands/foreground.rs
//! 前台检测 commands
//!
//! 对接 modules/foreground_inspection.rs，启动检测循环并把检测结果通过 Tauri 事件推送到前端

use std::path::PathBuf;
use std::sync::atomic::Ordering;

use serde::Serialize;
use tauri::{AppHandle, Emitter, Manager, State};
use tokio::sync::mpsc;

use crate::modules::foreground_inspection::{self, DetectionEvent, DetectionState};
use crate::state::AppState;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ForegroundStatus {
    pub running: bool,
    pub has_api_key: bool,
}

/// list_config.json 路径（与 data.json 同目录：app_data/PomoSolo/data/）
fn get_list_config_path(app: &AppHandle) -> PathBuf {
    let mut path = app.path().app_data_dir().unwrap_or_else(|_| {
        dirs::data_dir().unwrap_or_else(|| PathBuf::from("."))
    });
    path.push("PomoSolo");
    path.push("data");
    let _ = std::fs::create_dir_all(&path);
    path.join("list_config.json")
}

/// 确保名单配置已从磁盘加载到内存（首次调用时执行）。
///
/// 检测启动和名单管理命令都会先走这里：
/// 1. 把解析好的路径写入 DetectionState.config_path（供检测循环持久化历史记录）
/// 2. 若内存配置为空，加载磁盘上的已有配置（重启后恢复用户名单与历史）
async fn ensure_list_config_loaded(app: &AppHandle, state: &DetectionState) {
    let need_load = {
        let mut guard = state.config_path.lock().unwrap_or_else(|e| e.into_inner());
        if guard.is_none() {
            *guard = Some(get_list_config_path(app));
            true
        } else {
            false
        }
    };

    if need_load {
        let path = get_list_config_path(app);
        match foreground_inspection::load_list_config(&path) {
            Ok(cfg) => {
                let mut mem = state.list_config.lock().await;
                // 仅当内存为空时加载，避免覆盖运行期间新增的名单
                if mem.whitelist.is_empty() && mem.blacklist.is_empty() && mem.history.is_empty() {
                    *mem = cfg;
                }
            }
            Err(e) => {
                eprintln!("[ForegroundDetection] 加载名单配置失败: {}，使用默认空配置", e);
            }
        }
    }
}

/// 启动前台检测循环
#[tauri::command]
pub async fn foreground_start(
    app: AppHandle,
    state: State<'_, AppState>,
) -> Result<(), String> {
    // 先确保名单配置已加载（检测循环要用白/黑名单与历史记录）
    ensure_list_config_loaded(&app, &state.detection_state).await;

    if state.detection_state.running.load(Ordering::Relaxed) {
        return Ok(());
    }

    let det_state = state.detection_state.clone();
    let (tx, mut rx) = mpsc::unbounded_channel::<DetectionEvent>();

    // 模块内部会 spawn 一个 tokio 任务定期检测前台窗口
    foreground_inspection::start_detection(det_state.clone(), tx);

    // 发射 foreground-ready 事件（检测启动成功）
    // 事件名与前端 src/api/foreground.ts 中 FOREGROUND_EVENTS.ready 对齐
    let _ = app.emit("foreground-ready", serde_json::json!({}));

    // 发射 foreground-status 事件（running: true）
    let has_api_key = {
        let guard = state.detection_state.api_key.read().await;
        guard.as_ref().map(|k| !k.is_empty()).unwrap_or(false)
    };
    let _ = app.emit(
        "foreground-status",
        ForegroundStatus {
            running: true,
            has_api_key,
        },
    );

    // 转发检测结果到 webview
    // 事件名与前端 src/api/foreground.ts 中 FOREGROUND_EVENTS.entertainmentDetected 对齐：
    // 前端监听 'foreground-entertainment-detected'，payload 为 DetectionResult
    let app_clone = app.clone();
    tokio::spawn(async move {
        while let Some(event) = rx.recv().await {
            match event {
                DetectionEvent::Entertainment(result) => {
                    let _ = app_clone.emit("foreground-entertainment-detected", result);
                }
                DetectionEvent::ApiKeyInvalid => {
                    // API Key 已配置但无效（401/403），通知前端
                    let _ = app_clone.emit("foreground-api-key-invalid", serde_json::json!({}));
                }
                DetectionEvent::Error(msg) => {
                    let _ = app_clone.emit("foreground-error", msg);
                }
            }
        }
        // channel 关闭 → 检测循环退出；若确实已停止（而非被新一轮 start 抢占），
        // 发射 foreground-status 事件（running: false）
        if !det_state.running.load(Ordering::Relaxed) {
            let has_api_key = {
                let guard = det_state.api_key.read().await;
                guard.as_ref().map(|k| !k.is_empty()).unwrap_or(false)
            };
            let _ = app_clone.emit(
                "foreground-status",
                ForegroundStatus {
                    running: false,
                    has_api_key,
                },
            );
        }
    });

    Ok(())
}

/// 停止前台检测
#[tauri::command]
pub async fn foreground_stop(state: State<'_, AppState>) -> Result<(), String> {
    foreground_inspection::stop_detection(&state.detection_state);
    // running=false 后，检测循环会在下一次 tick 退出，channel 关闭，
    // 由 foreground_start 中 spawn 的转发任务负责发射 foreground-status(running: false)
    Ok(())
}

/// 获取检测状态
#[tauri::command]
pub async fn foreground_get_status(
    state: State<'_, AppState>,
) -> Result<ForegroundStatus, String> {
    let running = state
        .detection_state
        .running
        .load(Ordering::Relaxed);
    let has_api_key = state
        .detection_state
        .api_key
        .read()
        .await
        .as_ref()
        .map(|k| !k.is_empty())
        .unwrap_or(false);
    Ok(ForegroundStatus {
        running,
        has_api_key,
    })
}

/// 设置 DeepSeek API Key（传 None / 空字符串清空）
#[tauri::command]
pub async fn foreground_set_api_key(
    state: State<'_, AppState>,
    key: Option<String>,
) -> Result<(), String> {
    let normalized = match key {
        Some(k) if !k.is_empty() => Some(k),
        _ => None,
    };
    // 直接写入 RwLock，避免使用模块内只接受 String 的 set_api_key
    *state.detection_state.api_key.write().await = normalized;
    Ok(())
}

/// 检查检测是否就绪（API Key 已配置）
#[tauri::command]
pub async fn foreground_is_ready(state: State<'_, AppState>) -> Result<bool, String> {
    let guard = state.detection_state.api_key.read().await;
    Ok(guard.as_ref().map(|k| !k.is_empty()).unwrap_or(false))
}

// ===== 名单管理 commands（对应旧版 ipc-foreground.js 的 4 个名单频道） =====

/// 添加关键词到白名单
#[tauri::command]
pub async fn foreground_add_whitelist(
    app: AppHandle,
    state: State<'_, AppState>,
    keyword: String,
) -> Result<bool, String> {
    ensure_list_config_loaded(&app, &state.detection_state).await;
    Ok(foreground_inspection::add_to_whitelist(&state.detection_state, &keyword).await)
}

/// 添加关键词到黑名单
#[tauri::command]
pub async fn foreground_add_blacklist(
    app: AppHandle,
    state: State<'_, AppState>,
    keyword: String,
) -> Result<bool, String> {
    ensure_list_config_loaded(&app, &state.detection_state).await;
    Ok(foreground_inspection::add_to_blacklist(&state.detection_state, &keyword).await)
}

/// 将历史记录中的窗口标题标记为"不是娱乐"
///（前端 ForegroundWarning"不是娱乐"按钮：source 为 history/ai 时调用）
#[tauri::command]
pub async fn foreground_mark_history_not(
    app: AppHandle,
    state: State<'_, AppState>,
    window_title: String,
) -> Result<bool, String> {
    ensure_list_config_loaded(&app, &state.detection_state).await;
    Ok(foreground_inspection::mark_history_not(&state.detection_state, &window_title).await)
}

/// 把黑名单关键词移到白名单（误判纠正）
///（前端 ForegroundWarning"不是娱乐"按钮：source 为 blacklist 时调用）
#[tauri::command]
pub async fn foreground_move_blacklist_to_whitelist(
    app: AppHandle,
    state: State<'_, AppState>,
    keyword: String,
) -> Result<bool, String> {
    ensure_list_config_loaded(&app, &state.detection_state).await;
    Ok(foreground_inspection::move_blacklist_to_whitelist(&state.detection_state, &keyword).await)
}

/// 获取当前名单配置（用户白/黑名单 + 历史记录），供设置页展示
#[tauri::command]
pub async fn foreground_get_config(
    app: AppHandle,
    state: State<'_, AppState>,
) -> Result<foreground_inspection::ListConfig, String> {
    ensure_list_config_loaded(&app, &state.detection_state).await;
    let cfg = state.detection_state.list_config.lock().await;
    Ok(cfg.clone())
}
